### ISSUE TITLE
Spaces in URL (properly encoded) cause errors

### ISSUE THREAD

**Issue by BradyRosino** [bwssytems/ha-bridge#252]
With Indigo 7 you interact with the Rest API with the device names. Many of my device names have spaces in the names and it seems that HA Bridge does not care for this. Funny part is the action actually works, but HA Bridge thinks it failed. Google Home doesn't care about the error but Amazon Echo does, will always say the action failed.

This works -

http://localhost/devices/FamilyRoom_Light_Couch?isOn=1&_method=put

The following will result in an HA Bridge error, but the action actually does work (Alexa will just complain a lot saying she couldn't turn off the light when she really did)

http://localhost/devices/FamilyRoom%20-%20Light%20-%20Couch?isOn=1&_method=put

I'd love to see the spaces supported so that I don't have to rename all of my devices :-) Indigo does have the concept of a device ID to use in place of the names so renaming something doesn't break anything... but the RESTful API is a bit older and doesn't yet support, they've said they will be fixing at some point in the future. Currently only Python and AppleScript can use the ID.

Thanks for all the work on this bridge, really great stuff!

**Comment by BradyRosino** [comment]
Here is an example of one of the errors

`2016-11-19 10:20:30,383 [qtp322610576-22] WARN  com.bwssystems.HABridge.hue.HueMulator - Error calling out to HA gateway: IOException in log
org.apache.http.client.ClientProtocolException: null
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at com.bwssystems.HABridge.hue.HueMulator.doHttpRequest(HueMulator.java:1083)
	at com.bwssystems.HABridge.hue.HueMulator.lambda$19(HueMulator.java:941)
	at spark.RouteImpl$1.handle(RouteImpl.java:58)
	at spark.webserver.MatcherFilter.doFilter(MatcherFilter.java:162)
	at spark.webserver.JettyHandler.doHandle(JettyHandler.java:61)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:189)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
	at org.eclipse.jetty.server.Server.handle(Server.java:517)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:302)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:245)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.http.ProtocolException: Invalid redirect URI: http://localhost:8176/devices/Kitchen - Light - Table
	at org.apache.http.impl.client.DefaultRedirectStrategy.createLocationURI(DefaultRedirectStrategy.java:199)
	at org.apache.http.impl.client.DefaultRedirectStrategy.getLocationURI(DefaultRedirectStrategy.java:147)
	at org.apache.http.impl.client.DefaultRedirectStrategy.getRedirect(DefaultRedirectStrategy.java:220)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:121)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	... 23 common frames omitted
Caused by: java.net.URISyntaxException: Illegal character in path at index 37: http://localhost:8176/devices/Kitchen - Light - Table
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at org.apache.http.impl.client.DefaultRedirectStrategy.createLocationURI(DefaultRedirectStrategy.java:188)
	... 27 common frames omitted
2016-11-19 10:20:30,384 [qtp322610576-22] WARN  com.bwssystems.HABridge.hue.HueMulator - Error on calling url to change device state: http://localhost:8176/devices/Kitchen%20-%20Light%20-%20Table?isOn=1&_method=put`

**Comment by BradyRosino** [comment]
Looking closer in to the error, it looks like HA bridge is maybe not maintaining the URL encoding before making the call.

You'll notice the final HA Bridge error includes the %20 but everywhere else it actually just displays the space.

**Comment by BradyRosino** [comment]
Updated to 3.5.1 but it looks like this is still an issue. Let me know if I am missing something.

`2016-11-24 08:53:35,680 [qtp695017736-13] WARN  com.bwssystems.HABridge.hue.HueMulator - Error calling out to HA gateway: IOException in log
org.apache.http.client.ClientProtocolException: null
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at com.bwssystems.HABridge.hue.HueMulator.doHttpRequest(HueMulator.java:1173)
	at com.bwssystems.HABridge.hue.HueMulator.lambda$19(HueMulator.java:1024)
	at spark.RouteImpl$1.handle(RouteImpl.java:58)
	at spark.webserver.MatcherFilter.doFilter(MatcherFilter.java:162)
	at spark.webserver.JettyHandler.doHandle(JettyHandler.java:61)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:189)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
	at org.eclipse.jetty.server.Server.handle(Server.java:517)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:302)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:245)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.http.ProtocolException: Invalid redirect URI: http://localhost:8176/devices/Office Brady - Light - Desk
	at org.apache.http.impl.client.DefaultRedirectStrategy.createLocationURI(DefaultRedirectStrategy.java:199)
	at org.apache.http.impl.client.DefaultRedirectStrategy.getLocationURI(DefaultRedirectStrategy.java:147)
	at org.apache.http.impl.client.DefaultRedirectStrategy.getRedirect(DefaultRedirectStrategy.java:220)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:121)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	... 23 common frames omitted
Caused by: java.net.URISyntaxException: Illegal character in path at index 36: http://localhost:8176/devices/Office Brady - Light - Desk
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at org.apache.http.impl.client.DefaultRedirectStrategy.createLocationURI(DefaultRedirectStrategy.java:188)
	... 27 common frames omitted
2016-11-24 08:53:35,684 [qtp695017736-13] WARN  com.bwssystems.HABridge.hue.HueMulator - Error on calling url to change device state: http://localhost:8176/devices/Office%20Brady%20-%20Light%20-%20Desk?brightness=100&_method=put`

**Comment by audiofreak9** [comment]
Have you tried a plus sign `+` rather than the `%20`?

**Comment by melloware** [comment]
+1. I was going to ask about this too.  I think in all of the UI's where editing a URL is possible the HA Bridge should scrub the URL first using the built in Java URLEncoder like..

```java
String q = "random word £500 bank $";
String url = "http://example.com/query?q=" + URLEncoder.encode(q, "UTF-8");
```

This will turn spaces into "+" and all other characters to their % equivalent.  I got burned by this today as I needed to enter query param "A1 OFF" and I got a JSON exception until I manually encoded it.

Let me know if you want a PR fixing this?
